core/blockchain.go

```diff
@@ -811,6 +811,10 @@ func (bc *BlockChain) ExportN(w io.Writer, first uint64, last uint64) error {
 //
 // Note, this function assumes that the `mu` mutex is held!
 func (bc *BlockChain) writeHeadBlock(block *types.Block) {
+	if bc.Config().IsL2Migration(new(big.Int).Sub(block.Number(), big.NewInt(1))) {
```

Comment:
i feel there are better places to do this check, and return an error (vs. just log Crit which i'm not sure if it does not have the side effect of killing the process)

I remeber there are 2 entry point to add block here.

 1. As part of sync process. Blockchain.Insert()
 2. As part of consensus validators use another method that bypass the first pass of Blockchain, but it is a public method.

Reply:
I removed the log.Crit here, still looking into whether there's a better place to put the check

Reply:
There seem to be many code paths that can call [writeHeadBlock](https://github.com/celo-org/celo-blockchain/blob/02d8916d201dde648c4ea1650b7c3c9cfe5b8c89/core/blockchain.go#L813) including
[InsertPreprocessedBlock](https://github.com/celo-org/celo-blockchain/blob/02d8916d201dde648c4ea1650b7c3c9cfe5b8c89/core/blockchain.go#L1479)
[reorg](https://github.com/celo-org/celo-blockchain/blob/02d8916d201dde648c4ea1650b7c3c9cfe5b8c89/core/blockchain.go#L2206) (shouldn’t be relevant for Celo as I understand it)
[InsertChain](https://github.com/celo-org/celo-blockchain/blob/02d8916d201dde648c4ea1650b7c3c9cfe5b8c89/core/blockchain.go#L1714)
[InsertChainWithoutSealVerification](https://github.com/celo-org/celo-blockchain/blob/02d8916d201dde648c4ea1650b7c3c9cfe5b8c89/core/blockchain.go#L1746)
[insertSideChain](https://github.com/celo-org/celo-blockchain/blob/02d8916d201dde648c4ea1650b7c3c9cfe5b8c89/core/blockchain.go#L2063)
My reasoning behind calling bc.StopInsert() in writeHeadBlock is just that it seems to be called every time we add a new head block regardless of the code path and so feels safer than adding the check everywhere else. From your comment it seemed you may have been thinking [insertChain](https://github.com/celo-org/celo-blockchain/blob/02d8916d201dde648c4ea1650b7c3c9cfe5b8c89/core/blockchain.go#L1765) would be a good spot to add the check?

Reply:
I've added a test to check that InsertChain behaves as expected and returns an error when inserting large chains containing the migration block. 

Given that StopInsert() can be called multiple times I'm not worried about calling it in a private method, and this approach avoids changes in all the functions that call `writeHeadBlock`. Given I'm a bit new to the code, it feels safer to put the checks here so that it's easy to reason about when StopInsert() gets called, and so it gets called as early as possible after the last l1 block is added. 

There may be something I'm missing so please let me know if you still prefer a different spot for the checks!